pkg/receive/handler.go

```diff
@@ -747,82 +761,49 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 	// This is needed if context is canceled or if we reached success of fail quorum faster.
 	defer func() {
 		go func() {
-			for err := range ec {
-				if err != nil {
-					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
+			for wresp := range ec {
+				if wresp.err != nil {
+					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", wresp.err)
 				}
 			}
 		}()
 	}()
 
-	var success int
+	quorum := h.writeQuorum()
+	failureThreshold := h.options.ReplicationFactor - uint64(quorum)
+	successes := make([]int, numSeries)
+	seriesErrs := make([]errutil.MultiError, numSeries)
 	for {
 		select {
 		case <-fctx.Done():
 			return fctx.Err()
-		case err, more := <-ec:
+		case wresp, more := <-ec:
 			if !more {
+				for _, rerr := range seriesErrs {
+					if seriesReplicated {
+						errs.Add(rerr.Err())
```

Comment:
This might become verbose because we're adding an error for each series inside a write request.

Reply:
Agreed but I'm not sure how would that make it possible for us to determine the final error?

Reply:
I've added an `Error()` method that deduplicates reasons. This is mostly for for better readability of what has gone wrong